ceremonies/rpc/src/lib.rs

```diff
@@ -38,26 +45,90 @@ where
 	) -> Result<Vec<(CeremonyIndexType, CommunityReputation)>>;
 }
 
-pub struct Ceremonies<Client, Block, AccountId> {
+pub struct Ceremonies<Client, Block, AccountId: ByteArray, S> {
 	client: Arc<Client>,
 	_marker: std::marker::PhantomData<(Block, AccountId)>,
 	deny_unsafe: DenyUnsafe,
+	storage: Arc<RwLock<S>>,
+	offchain_indexing: bool,
 }
 
-impl<Client, Block, AccountId> Ceremonies<Client, Block, AccountId> {
+impl<Client, Block, AccountId, S> Ceremonies<Client, Block, AccountId, S>
+where
+	S: 'static + OffchainStorage,
+	Block: sp_api::BlockT,
+	AccountId: 'static + Encode + Decode + Send + Sync + ByteArray,
+	Client: Send + Sync + 'static + ProvideRuntimeApi<Block> + HeaderBackend<Block>,
+	Client::Api: CeremoniesRuntimeApi<Block, AccountId>,
+{
 	/// Create new `Ceremonies` instance with the given reference to the client.
-	pub fn new(client: Arc<Client>, deny_unsafe: DenyUnsafe) -> Self {
-		Ceremonies { client, _marker: Default::default(), deny_unsafe }
+	pub fn new(
+		client: Arc<Client>,
+		deny_unsafe: DenyUnsafe,
+		storage: S,
+		offchain_indexing: bool,
+	) -> Self {
+		Ceremonies {
+			client,
+			_marker: Default::default(),
+			deny_unsafe,
+			storage: Arc::new(RwLock::new(storage)),
+			offchain_indexing,
+		}
+	}
+
+	/// Check if cache was marked dirty by the runtime
+	pub fn cache_dirty(&self, key: &[u8]) -> bool {
+		match self.storage.read().get(STORAGE_PREFIX, key) {
+			Some(d) => Decode::decode(&mut d.as_slice()).unwrap_or_else(|e| {
+				log::error!("Cache dirty bit: {:?}", e);
+				log::info!("{:?}: Defaulting to dirty == true", key);
+				true
+			}),
+			None => {
+				log::warn!("{:?}: Cache dirty bit is none.", key);
+				true
+			},
+		}
+	}
+
+	pub fn get_storage<V: Decode>(&self, key: &[u8]) -> Option<V> {
+		match self.storage.read().get(STORAGE_PREFIX, key) {
+			Some(v) => Some(Decode::decode(&mut v.as_slice()).unwrap()),
+			None => None,
+		}
+	}
+
+	pub fn set_storage<V: Encode>(&self, key: &[u8], val: &V) {
+		self.storage.write().set(STORAGE_PREFIX, key, &val.encode());
+	}
+
+	pub fn refresh_reputation_cache(
+		&self,
+		account: AccountId,
+		at: Option<<Block as BlockT>::Hash>,
+	) {
+		let api = self.client.runtime_api();
+		let at = BlockId::hash(at.unwrap_or_else(|| self.client.info().best_hash));
+		let reputations =
+			api.get_reputations(&at, &account).map_err(runtime_error_into_rpc_err).unwrap();
+		let cache_key = &(STORAGE_REPUTATION_KEY, account.as_slice()).encode()[..];
```

Comment:
Actually, the cleanest approach would be to define a function in the primitives that derives the cache key because it is used multiple times, and it is very easy to introduce discrepancies.

```
pub fn reputation_cache_key<Account: Encode>(account: Account) -> Vec<u8> {
      (STORAGE_REPUTATION_KEY, account).encode()
}
```

Reply:
Please to the same for the cache dirty bit.

Reply:
when you explain it like this, it sounds so easy :) but it somehow isn't😅

Reply:
haha, it took me a while to get there :)